include/seastar/coroutine/generator.hh

```diff
@@ -23,525 +23,438 @@
-concept Fifo = requires(Container<T>&& c, T&& value) {
-    // better off returning a reference though, so we can move away from it
-    { c.front() } -> std::same_as<T&>;
-    c.pop_front();
-    c.push_back(value);
-    bool(c.empty());
-    { c.size() } -> std::convertible_to<size_t>;
-};
+template <typename Yielded> class next_awaiter;
 
-template<typename T>
-concept NothrowMoveConstructible = std::is_nothrow_move_constructible_v<T>;
+template <typename Yielded>
+class generator_promise_base : public seastar::task {
+protected:
+    // a glvalue yield expression is passed to co_yield as its operand. and
+    // the object denoted by this expression is guaranteed to live until the
+    // coroutine resumes. we take advantage of this fact by storing only a
+    // pointer to the denoted object in the promise as long as the result of
+    // dereferencing this pointer is convertible to the Ref type.
+    std::add_pointer_t<Yielded> _value = nullptr;
 
-template<NothrowMoveConstructible T>
-class generator_unbuffered_promise final : public seastar::task {
-    using generator_type = seastar::coroutine::experimental::generator<T, std::optional>;
-    std::optional<seastar::promise<>> _wait_for_next_value;
-    generator_type* _generator = nullptr;
+protected:
+    std::exception_ptr _exception;
+    std::coroutine_handle<> _consumer;
+    task* _waiting_task = nullptr;
+
+    /// awaiter returned by the generator when it produces a new element
+    ///
+    /// There are different combinations of expression types passed to
+    /// \c co_yield and \c Ref. In most cases, zero copyies are made. Copies
+    /// are only necessary when \c co_yield requires type conversion.
+    ///
+    /// The following table summarizes the number of copies made for different
+    /// scenarios:
+    ///
+    /// | Ref       | co_yield const T& | co_yield T& | co_yield T&& | co_yield U&& |
+    /// | --------- | ----------------- | ----------- | ------------ | ------------ |
+    /// | T         | 0                 | 0           | 0            | 1            |
+    /// | const T&  | 0                 | 0           | 0            | 1            |
+    /// | T&        | ill-formed        | 0           | ill-formed   | ill-formed   |
+    /// | T&&       | ill-formed        | ill-formed  | 0            | 1            |
+    /// | const T&& | ill-formed        | ill-formed  | 0            | 1            |
+    ///
+    /// When no copies are required, \c yield_awaiter is used. Otherwise,
+    /// \c copy_awaiter is used. The latter converts \c U to \c T, and keeps the converted
+    /// value in it.
+    class yield_awaiter final {
+        generator_promise_base* _promise;
+        std::coroutine_handle<> _consumer;
+    public:
+        yield_awaiter(generator_promise_base* promise,
+                      std::coroutine_handle<> consumer) noexcept
+            : _promise{promise}
+            , _consumer{consumer}
```

Comment:
nit: why the curly brances?
We usually use parenthesis for member initialization.

Reply:
i think in addition to the reasons listed by https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-list , the curly braces are visually more clear that we are initializing variables instead of calling functions.

Reply:
FWIW, we also use curly braces when initializing members in-class , see https://github.com/scylladb/seastar/blob/e42e382648e459a608bf42f9d8c1d95550182362/include/seastar/core/reactor.hh#L317-L330 . but yeah, this might be different from the initializer list . but i think the idea is the same.